mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOps.td

```diff
@@ -884,7 +885,8 @@ def SYCLItemGetLinearIDOp
 
 def SYCLNDItemGetGlobalIDOp
     : SYCLMethodOpInterfaceImpl<"nd_item.get_global_id", "NdItemType",
-                                ["get_global_id"], [Pure, SYCLGetID]> {
+                                ["get_global_id"], 
```

Comment:
While I think your patch is correct overall, I'm not too sure about that. By spec this is "pure" as the `nd_item` is immutable and may not even read memory (this function really maps to `SYCLGlobalIDOp`). This applies to `nd_item`, `item`, `group` and `sub_group` but not `id`, `range` and `nd_range`.

Reply:
> `nd_range`

Isn't `nd_range` also immutable?

Reply:
I see. If I'm correct, thanks to your LICM pass, we will hoist even if the operation is not pure, right?

Reply:
Correct, as long as there are no operation in the loop that store to the memory that is read by another operation (actually is the store dominates the load we can still hoist both operations).